storagenode/contact/chore.go

```diff
@@ -48,60 +58,111 @@ func (chore *Chore) Run(ctx context.Context) (err error) {
 	defer mon.Task()(&ctx)(&err)
 	chore.log.Info("Storagenode contact chore starting up")
 
+	var group errgroup.Group
+
 	if !chore.service.initialized.Wait(ctx) {
 		return ctx.Err()
 	}
 
-	return chore.Loop.Run(ctx, func(ctx context.Context) error {
-		if err := chore.pingSatellites(ctx); err != nil {
-			chore.log.Error("pingSatellites failed", zap.Error(err))
-		}
-		return nil
-	})
+	initialBackOff := time.Second
+	interval := chore.interval
+	chore.mu.Lock()
+	for _, satellite := range chore.trust.GetSatellites(ctx) {
+		satellite := satellite
+
+		cycle := sync2.NewCycle(interval)
+		chore.cycles = append(chore.cycles, cycle)
+
+		cycle.Start(ctx, &group, func(ctx context.Context) error {
+			chore.log.Info("contact cycle: satellite " + satellite.String() + " - starting contact cycle")
+			interval := initialBackOff
+			attempts := 0
+			for {
+				err := chore.pingSatellite(ctx, satellite)
+				attempts++
+				if err == nil {
+					return nil
+				}
+				chore.log.Error("contact cycle: satellite " + satellite.String() + " - ping satellite failed " + strconv.Itoa(attempts) + " times - " + err.Error())
+
+				// Sleeps until interval times out, then continue. Returns if context is cancelled.
+				if !sync2.Sleep(ctx, interval) {
+					chore.log.Info("contact cycle: satellite " + satellite.String() + " - contact chore context cancelled")
+					return nil
+				}
+				if interval == chore.interval {
+					chore.log.Error("contact cycle: satellite " + satellite.String() + " - retries timed out for this cycle")
+					return nil
+				}
+				interval *= 2
+				if interval > chore.interval {
+					interval = chore.interval
+				}
+			}
+		})
+	}
+	chore.mu.Unlock()
+	chore.started.Release()
+	return group.Wait()
 }
 
-func (chore *Chore) pingSatellites(ctx context.Context) (err error) {
+func (chore *Chore) pingSatellite(ctx context.Context, id storj.NodeID) (err error) {
 	defer mon.Task()(&ctx)(&err)
-	var group errgroup.Group
 	self := chore.service.Local()
-	satellites := chore.trust.GetSatellites(ctx)
+	address, err := chore.trust.GetAddress(ctx, id)
+	if err != nil {
+		return errPingSatellite.New("failed to get satellite address %s: %v", id.String(), err)
```

Comment:
I'm wondering if we should keep the stack trace of `err` using `Wrap` but then we'll lose the message.

Reply:
yeah that will probably be more useful. I didn't know that .New() stripped out the stack trace

Reply:
`New` will create a stack trace but from this point. It doesn't consider the stack trace of `err` because it isn't taking `err` as a parameter as `Wrap` does.

There is an alternative but I dislike it. The alternative is to add the stack trace to the message, using the `+` to the used verb (e.g. `+v`).

Reply:
gotcha, yeah that makes sense. I don't think the messages are crucial if we have the stack trace, so I will use wrap